Add lint and fix .PHONY in Makefile

sambooo opened the pull request:
Adds `make lint` which can be called in CI to fail when there are lint issues, as described in #456. Travis should fail here with a linting error after passing the test suite.

Credit to the [Hugo](https://github.com/spf13/hugo) team for the code, not really my own work

jdoliner commented:
@sambooo thanks for another pull!!

CI is failing for good reasons here. This PR adds linter to CI but the linter fails because we don't have comments on all our exported classes. I'm going to add some commits of my own to this PR to get the linter fixed since I don't think @sambooo should be expected to document all of our undocumented things as part of this PR.

jdoliner commented:
Updates on this. Turns out I can't add commits directly to this PR since the branch is on @sambooo's gh account not mine. I pulled it in to my own branch and fixed the immediate linter issues with `src/client`. Unfortunately there's a snag `pb.go` files don't pass the linter. I think what we want to do is use `go list` little more creatively to give us file names instead of pkg names and the `grep -v pb.go` to get rid of the stuff we don't want to lint.

sambooo commented:
Fixed some conflicts and removed the change that caused travis to fail so `make lint` can be merged. I'll open another PR with the failing travis command.

jdoliner commented:
Cool, this all lgtm now.
